Add integration tests for RBAC v2

pitlv2109 opened the pull request:
* Add integration tests for RBAC v2
* Add Galley when create testing apps for RBAC v2 and existing security tests.

Will add more tests for RBAC v2 later to cover global/local namespaces, inline role def, etc.

For https://github.com/istio/istio/issues/12394
Fix https://github.com/istio/istio/issues/13352

istio-testing commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio/pull/13353#" title="Author self-approved">pitlv2109</a>*
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **nmittler**

If they are not already assigned, you can assign the PR to them by writing `/assign @nmittler` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[pkg/OWNERS](https://github.com/istio/istio/blob/master/pkg/OWNERS)**
- **[tests/integration/OWNERS](https://github.com/istio/istio/blob/master/tests/integration/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["nmittler"]} -->

istio-testing commented:
@pitlv2109: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-multicluster-e2e.sh | 4a5290f71a0b3f9a13409ebd6923fd4d7addd01a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13353/istio-pilot-multicluster-e2e/11077) | `/test istio-pilot-multicluster-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

pitlv2109 commented:
@yangminzhu @liminw This PR is ready for review for RBAC v2 integration tests.
@nmittler I added `path` and `AppProtocolTCP` to the test framework. Let me know if you rather just reuse `AppProtocolHTTP` for `tcp`. Also fixed Galley being nil as we discussed.

yangminzhu commented on pkg/test/util/policy/policy.go:
```diff
@@ -43,9 +43,16 @@ func (p TestPolicy) TearDown() {
+}
+
+// ApplyPolicyAnyFilePath runs `kubectl -f <namespace> <fileName>` where fileName is the real/explicit path of the file.
+func ApplyPolicyAnyFilePath(t *testing.T, env *kube.Environment, namespace string, fileName string) *TestPolicy {
```
What does `Any` mean here?

pitlv2109 replied:
`ApplyPolicyFile` in `policy.go` assumes the yaml configs are under `testdata`. Since our RBAC yaml files using template for namespaces, it's better to put them in the tmp directory created by the test. So `AnyFilePath` means the file can be anywhere. I can change it to `ApplyPolicyAbsoluteFilePath` to make it more clear, wdyt?

yangminzhu replied:
Okay, I think `ApplyPolicyFromFilePath` might be better in this case.

yangminzhu commented on tests/integration/security/rbac/rbac_v2_test.go:
```diff
@@ -0,0 +1,147 @@
+	if cfg == nil {
+		return
+	}
+	//cfg.Values["global.mtls.enabled"] = "true"
```
Remove this line.

yangminzhu commented on tests/integration/security/rbac/rbac_v2_test.go:
```diff
@@ -0,0 +1,147 @@
+	} else {
+		successIfAuthEnabled = false
+	}
+	cfg.Values["sidecarInjectorWebhook.rewriteAppHTTPProbe"] = "true"
```
Why do we need this in RBAC test? If this is something common to all tests, probably better do it in the framework.

yangminzhu commented on tests/integration/security/rbac/rbac_v2_test.go:
```diff
@@ -0,0 +1,147 @@
+	connections := []connection.Connection{
+		// Port 80 is where HTTP is served, 90 is where TCP is served.
+		// In RBAC, ExpectedSuccess = true means that the client receives a valid, non-RBAC denied response from the server.
+		{To: appA, From: appB, Port: 80, Protocol: apps.AppProtocolHTTP, Path: "/xyz", ExpectedSuccess: false},
```
Can we check it's actually denied by RBAC filter when it's denied? The RBAC v1 test will check the response code is "403". This is important so that we know for sure it's denied by our RBAC filter, not by some random failures.

pitlv2109 replied:
Makes sense. Will do.

yangminzhu commented on tests/integration/security/rbac/rbac_v2_test.go:
```diff
@@ -0,0 +1,147 @@
+
+	// Sleep 5 seconds for the policy to take effect.
+	// TODO(pitlv2109: Check to make sure policies have been created instead.
+	time.Sleep(5 * time.Second)
```
Instead of waiting for an uncertain duration, it might be better to send each request 100 times and calculate the count of each response result, if 90% of the response matches the `ExpectedSuccess`, then we consider it passed, otherwise it failed.

pitlv2109 replied:
I will add some code to actually see if the policies are there in a future PR.

yangminzhu commented on tests/integration/security/rbac/testdata/istio-rbac-v2-rules.yaml.tmpl:
```diff
@@ -0,0 +1,129 @@
+# istio-rbac-v2-rules.yaml to enforce access control for both http and tcp services using Istio RBAC v2 rules.
```
We will need a test case for namespace authz policy, which means the workload_selector is not specified and there is also no destination.labels in ServiceRole. We can do this in follow-up PRs.

pitlv2109 replied:
Yes. We can add more tests in follow-up PRs. This is just to get the basic tests in.

istio-testing commented:
@pitlv2109: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-multicluster-e2e.sh | 4a5290f71a0b3f9a13409ebd6923fd4d7addd01a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13353/istio-pilot-multicluster-e2e/11077) | `/test istio-pilot-multicluster-e2e`
prow/istio-integ-k8s-tests.sh | 4a5290f71a0b3f9a13409ebd6923fd4d7addd01a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13353/istio-integ-k8s-tests/6824) | `/test istio-integ-k8s-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

nmittler commented:
/lgtm

... for the test fixes that add Galley.

istio-testing commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio/pull/13353#issuecomment-483716183" title="LGTM">nmittler</a>*, *<a href="https://github.com/istio/istio/pull/13353#" title="Author self-approved">pitlv2109</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[pkg/OWNERS](https://github.com/istio/istio/blob/master/pkg/OWNERS)~~ [nmittler]
- ~~[tests/integration/OWNERS](https://github.com/istio/istio/blob/master/tests/integration/OWNERS)~~ [nmittler]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

istio-testing commented:
@pitlv2109: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-integ-k8s-tests.sh | 4a5290f71a0b3f9a13409ebd6923fd4d7addd01a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13353/istio-integ-k8s-tests/6824) | `/test istio-integ-k8s-tests`
prow/istio-pilot-multicluster-e2e.sh | 4a5290f71a0b3f9a13409ebd6923fd4d7addd01a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13353/istio-pilot-multicluster-e2e/11095) | `/test istio-pilot-multicluster-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

howardjohn commented:
This shouldn't have been merged. The test fails and @nmittler approval was only for a part of it @yangminzhu had open comments (No fault of anyone involved.. blame the bot). 

@pitlv2109 probably best to just address the comments in a followup pr?

pitlv2109 commented:
Yep. I'm resolving Yangmin's comments in a new PR now

istio-testing commented:
@pitlv2109: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-multicluster-e2e.sh | 4a5290f71a0b3f9a13409ebd6923fd4d7addd01a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13353/istio-pilot-multicluster-e2e/11095) | `/test istio-pilot-multicluster-e2e`
prow/istio-integ-k8s-tests.sh | 4a5290f71a0b3f9a13409ebd6923fd4d7addd01a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13353/istio-integ-k8s-tests/6848) | `/test istio-integ-k8s-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

pitlv2109 commented:
@howardjohn I tested `TestSdsCitadelCaFlow` and `TestRBACV2` yesterday and they both passed. However many tests are failing today. I reverted back to the checked in PR before mine and still experiencing the same issue. Seems like tests are timed out because testing apps are having issues with error `Internal error occurred: admission webhook "sidecar-injector.istio.io" denied the request: error converting YAML to JSON: yaml: line 3: could not find expected ':'`
